### DISCUSSION TITLE
auth::attempt() error Undefined array key "password"

### DISCUSSION THREAD

**Discussion by mlaicadmin** [livewire/livewire#7232] [1. Help]
I am trying to create a manuel login form. It recognizes when data is not in the database, but whenever the login is successful, I get a weird Undefined array key "password" error. What's strange is that I am only validating based upon the First Name and Last Name. 
I also got the majority of code from videos on manuel authentication, with others being able to login with the same exact code. I am not sure as to why I am getting this error.

Here is my view:
```
@if (session()->has("message"))
    <div class="p-10 text-2xl">
        {{ session("message") }}
    </div>
    @endif

    <form wire:submit.prevent="loginUser" class="w-full max-w-lg">
        <div class="flex flex-wrap -mx-3 mb-6">
          <div class="w-full md:w-1/2 px-3 mb-6 md:mb-0">
            <label class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2" for="grid-first-name">
              First Name
            </label>
            <input name='First_Name' wire:model='First_Name' class="appearance-none block w-full bg-gray-200 text-gray-700 border-gray-200 rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white" id="grid-first-name" type="text" placeholder="John">
            @error('First_Name') <span class="text-danger error">{{ $message }}</span>@enderror
          </div>
          <div class="w-full md:w-1/2 px-3">
            <label class="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2" for="grid-last-name">
              Last Name
            </label>
            <input name='Last_Name' wire:model='Last_Name' class="appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500" id="grid-last-name" type="text" placeholder="Doe">
            @error('Last_Name') <span class="text-danger error">{{ $message }}</span>@enderror
          </div>
        </div>
        <div class="inline-flex">
            <button type="submit" class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
              Login
            </button>
          </div> 
      </form>
```

Here is my controller:
```
<?php

namespace App\Http\Livewire\Layout;

use Livewire\Component;
use Illuminate\Support\Facades\Auth;
use App\Models\NewUserStudent_Authenticatable;
use Illuminate\Http\Request;
use Illuminate\Foundation\Auth\User as Authenticatable;

class Login extends Component {

    public $Email, $Password, $First_Name, $Last_Name;



    public function loginUser() 
    {

       $this->validate([
            'First_Name' => "required|max:255",
            'Last_Name' => "required|max:255"
        ]);

        // check if password and email matches
            if(Auth::attempt(array('First_Name' => $this->First_Name, 'Last_Name' => $this->Last_Name))) 
            {
         

                session()->flash("success", 'You are logged in!');


                // redirect to home page which allows user to see logout button
                // return $this->redirect('Home');
            } else {

            // if user not correct
            $this->addError("First_Name", 'The password or email is not valid'); 
            }
    }

    public function render()
    {
        return view('livewire.layout.login');
    }

}

```

**Comment by CyberPunkCodes** [comment]
This is a Laravel issue not a Livewire one.

Your problem is `Auth::attempt()` requires a key named `password`. Lowercased just like that and not `Password`. All of the remaining keys passed in credentials will be used to find the user by being combined to the where statement. Those keys need to match columns in your database, including the casing.

```php
attempt(array $credentials = [], bool $remember = false)
```

Which, the credentials gets passed through:

```php
public function retrieveByCredentials(array $credentials)
{
    $credentials = array_filter(
        $credentials,
        fn ($key) => ! str_contains($key, 'password'),
        ARRAY_FILTER_USE_KEY
    );

    if (empty($credentials)) {
        return;
    }
    
    ...
}
```

`str_contains` is case sensitive!

You should really use a unique identifier for the rest of the credentials though, like username or email address. If you have more than one user with the same values, it will cause problems.

99% of PHP code, especially laravel, the keys are all lowercased. That is why naming conventions and PSRs were created, so everyone is on (roughly) the same page.

So we would user either `firstName` or `first_name` and not `First_Name`. While I don't like to critique other people's code, following the traditional naming conventions can save you a lot of trouble in the future.

Typically `camelCase` is for variables, `PascalCase` is for classes, and `snake_case` is for databases and columns. In certain cases, laravel can handle things so long as they follow supported naming conventions. So it can prevent future errors for yourself. Not to mention if you try to submit a PR to someone else's project, it may get rejected solely based on using something like`$First_Name`. For example, laravel is smart with handling relations. If you name your database columns in a way that it likes, ie: `user_id`, it can match the relation without having to pass additional values to tell it exactly which column to look for.

You will see on the Laravel docs, under the [Contributing](https://laravel.com/docs/10.x/contributions) section, that Laravel uses the PSR-2 coding style. It is worth mentioning that the PSR-2 style is deprecated and we should use PSR-12. They are almost identical and I don't think you will have any issues following 12 if you were to, in the future, submit a Pull Request to the Laravel repo or someone else's package. I am not sure why Laravel still technically follows 2 when it's been deprecated since 2019.

While you don't have to personally follow it to a T, or have a linter running in your IDE to keep you in check, it's worth looking over the PSR coding standard and applying it. Naming conventions, 4 spaces and not a tab, etc. Though, having a linter running in your IDE for a while will help mold you into following it naturally until it's just the way you code.
